core/chains/evm/headtracker/heads.go

```diff
@@ -26,114 +27,158 @@ type Heads interface {
 // Trims old blocks whose height is smaller than minBlockToKeep
 func (h *heads) MarkFinalized(finalized common.Hash, minBlockToKeep int64) bool {
 	h.mu.Lock()
 	defer h.mu.Unlock()
 
-	if len(h.heads) == 0 {
+	if len(h.headsByHash) == 0 {
 		return false
 	}
 
-	// deep copy to avoid race on head.Parent
-	h.heads, h.headsMap = deepCopy(h.heads, minBlockToKeep)
-
-	finalizedHead, ok := h.headsMap[finalized]
+	finalizedHead, ok := h.headsByHash[finalized]
 	if !ok {
 		return false
 	}
-	for finalizedHead != nil {
-		finalizedHead.IsFinalized = true
-		finalizedHead = finalizedHead.Parent
+
+	markFinalized(finalizedHead)
+
+	// remove all blocks that are older than minBlockToKeep
+	for h.headsAsc.Len() > 0 && h.headsAsc.Peek().Number < minBlockToKeep {
+		oldBlock := heap.Pop(h.headsAsc).(*evmtypes.Head)
+		delete(h.headsByHash, oldBlock.Hash)
+		// clear .Parent in oldBlock's children
+		for _, oldBlockChildren := range h.headsByParent[oldBlock.Hash] {
+			oldBlockChildren.Parent.Store(nil)
+		}
+		// headsByParent are expected to be of the same height, so we can remove them all at once
+		delete(h.headsByParent, oldBlock.ParentHash)
+	}
+
+	if h.highest.Number < minBlockToKeep {
+		h.highest = nil
 	}
 
 	return true
 }
 
-func deepCopy(oldHeads []*evmtypes.Head, minBlockToKeep int64) ([]*evmtypes.Head, map[common.Hash]*evmtypes.Head) {
-	headsMap := make(map[common.Hash]*evmtypes.Head, len(oldHeads))
-	heads := make([]*evmtypes.Head, 0, len(headsMap))
-	for _, head := range oldHeads {
-		if head.Hash == head.ParentHash {
-			// shouldn't happen but it is untrusted input
-			continue
-		}
-		if head.BlockNumber() < minBlockToKeep {
-			// trim redundant blocks
-			continue
-		}
-		// copy all head objects to avoid races when a previous head chain is used
-		// elsewhere (since we mutate Parent here)
-		headCopy := *head
-		headCopy.Parent = nil // always build it from scratch in case it points to a head too old to be included
-		// map eliminates duplicates
-		// prefer head that was already in heads as it might have been marked as finalized on previous run
-		if _, ok := headsMap[head.Hash]; !ok {
-			headsMap[head.Hash] = &headCopy
-			heads = append(heads, &headCopy)
+func markFinalized(head *evmtypes.Head) {
+	// we can assume that if a head was previously marked as finalized all its ancestors were marked as finalized
+	for head != nil && !head.IsFinalized.Load() {
+		head.IsFinalized.Store(true)
+		head = head.Parent.Load()
+	}
+}
+
+func (h *heads) ensureNoCycles(newHead *evmtypes.Head) error {
+	if newHead.ParentHash == newHead.Hash {
+		return fmt.Errorf("cycle detected: newHeads reference itself newHead(%s)", newHead.String())
+	}
+	if parent, ok := h.headsByHash[newHead.ParentHash]; ok {
+		if parent.Number >= newHead.Number {
+			return fmt.Errorf("potential cycle detected while adding newHead as child: %w", newPotentialCycleError(parent, newHead))
```

Comment:
I understand this is a cycle problem in the context of Head Tracker, but this is more of a general problem as the parent's block number is higher than the child that references it. I would consider this as invalid head/data from the chain.

Reply:
`newPotentialCycleError` already signals that we expect block number `to strictly decrease in 'child -> parent' relation`.  Do you suggest changing this msg?

Reply:
`newPotentialCycleError` error message is accurate. Perhaps I would rephrase the `potential cycle...` part because if the hashes are right, from a user perspective the issue would be better categorized as bad data on the block number rather than a cycle problem, but it's just me nitpicking. You can ignore this.